### PR TITLE
helm: Enrich serviceMonitor properties

### PR DESCRIPTION
**Pull request by jkroepke**
Required for custom relabeling actions or istio deployments.

### CODE REVIEW

**Comment by pierluigilenoci** [comment]
@jkroepke, could you please add the ArtifactHub annotation here? 
https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/Chart.yaml#L35-L41

There are some missing default values here:
https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/values.yaml#L362-L374

Specifically:
```
scheme
bearerTokenFile
tlsConfig
```
And an unused one, `targetLabels`.

So please fix your PR.

File: helm/oauth2-proxy/values.yaml [hunk 1]
```diff
@@ -373,6 +373,29 @@ metrics:
...
+    relabelings: []
+    # - sourceLabels: [__meta_kubernetes_pod_node_name]
+    #   separator: ;
+    #   regex: ^(.*)$
+    #   targetLabel: nodename
+    #   replacement: $1
+    #   action: replace
+
+    targetLabels: []
+
```
Comment (jkroepke) [inline helm/oauth2-proxy/values.yaml:398]: @desaintmartin I have defined default values?

Comment (pierluigilenoci) [reply helm/oauth2-proxy/values.yaml:398]:
```
scheme
bearerTokenFile
tlsConfig
``` 
are missing.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 2]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    {{- end }}
+    {{- with .Values.metrics.serviceMonitor.scheme }}
+    scheme: {{ . }}
+    {{- end }}
+    {{- with .Values.metrics.serviceMonitor.bearerTokenFile }}
+    bearerTokenFile: {{ . }}
+    {{- end }}
+    {{- with .Values.metrics.serviceMonitor.tlsConfig }}
+    tlsConfig:
+      {{- toYaml .| nindent 8 }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:47]: This should be 4, not 8.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 3]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    {{- end }}
+    {{- with .Values.metrics.serviceMonitor.metricRelabelings }}
+    metricRelabelings:
+      {{- toYaml . | nindent 8 }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:51]: This should be 4, not 8.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 4]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    {{- end }}
+    {{- with .Values.metrics.serviceMonitor.relabelings }}
+    relabelings:
+      {{- toYaml . | nindent 8 }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:55]: This should be 4, not 8.

**Comment by pierluigilenoci** [comment]
Documentation should also be updated with the changes.
https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/README.md?plain=1#L204-L209

I apologize for the fragmented review. 
To speed up, I'm making suggestions while busy with other tasks. 
I hope I can be of help anyway.

**Comment by jkroepke** [comment]
I hope I hit all your points

File: helm/oauth2-proxy/values.yaml [hunk 5]
```diff
@@ -373,6 +373,40 @@ metrics:
...
+    relabelings: []
+    # - sourceLabels: [__meta_kubernetes_pod_node_name]
+    #   separator: ;
+    #   regex: ^(.*)$
+    #   targetLabel: nodename
+    #   replacement: $1
+    #   action: replace
+
+    ## Attach labels to the ServiceMonitor resource based on Service metadata.
+    targetLabels: []
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/values.yaml:408]: Unused, why is it here?

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 6]
```diff
@@ -2,6 +2,10 @@
 apiVersion: monitoring.coreos.com/v1
 kind: ServiceMonitor
 metadata:
+  {{- with .Values.metrics.servicemonitor.annotations }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:5]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 7]
```diff
@@ -26,6 +30,28 @@ spec:
   endpoints:
   - port: metrics
     path: "/metrics"
-    interval: {{ .Values.metrics.serviceMonitor.interval }}
-    scrapeTimeout: {{ .Values.metrics.serviceMonitor.scrapeTimeout }}
+    {{- with .Values.metrics.servicemonitor.interval }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:33]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 8]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    interval: {{ . }}
+    {{- end }}
+    {{- with .Values.metrics.servicemonitor.scrapeTimeout }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:36]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 9]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    scrapeTimeout: {{ . }}
+    {{- end }}
+    {{- with .Values.metrics.servicemonitor.scheme }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:39]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 10]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    scheme: {{ . }}
+    {{- end }}
+    {{- with .Values.metrics.servicemonitor.bearerTokenFile }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:42]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 11]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    bearerTokenFile: {{ . }}
+    {{- end }}
+    {{- with .Values.metrics.servicemonitor.tlsConfig }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:45]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 12]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    tlsConfig:
+      {{- toYaml .| nindent 4 }}
+    {{- end }}
+    {{- with .Values.metrics.servicemonitor.metricRelabelings }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:49]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/templates/servicemonitor.yaml [hunk 13]
```diff
@@ -26,6 +30,28 @@ spec:
...
+    metricRelabelings:
+      {{- toYaml . | nindent 4 }}
+    {{- end }}
+    {{- with .Values.metrics.servicemonitor.relabelings }}
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/templates/servicemonitor.yaml:53]: Please use the camel case `serviceMonitor`.

File: helm/oauth2-proxy/README.md [hunk 14]
```diff
@@ -207,6 +207,13 @@ Parameter | Description | Default
 `metrics.serviceMonitor.interval` | Prometheus scrape interval | `60s`
 `metrics.serviceMonitor.scrapeTimeout` | Prometheus scrape timeout | `30s`
 `metrics.serviceMonitor.labels` | Add custom labels to the ServiceMonitor resource| `{}`
+`metrics.serviceMonitor.scheme` | HTTP scheme to use for scraping. Can be used with `tlsConfig` for example if using istio mTLS.| `""`
+`metrics.serviceMonitor.tlsConfig` | TLS configuration to use when scraping the endpoint. For example if using istio mTLS.| `{}`
+`metrics.serviceMonitor.bearerTokenFile` | Path to bearer token file.| `""`
+`metrics.serviceMonitor.annotations` | Used to pass annotations that are used by the Prometheus installed in your cluster| `{}`
+`metrics.serviceMonitor.metricRelabelings` | Metric relabel configs to apply to samples before ingestion.| `[]`
+`metrics.serviceMonitor.relabelings` | Relabel configs to apply to samples before ingestion.| `[]`
+`metrics.serviceMonitor.targetLabels` | Attach labels to the ServiceMonitor resource based on Service metadata.| `[]`
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/README.md:216]: Unused, why is it here?

File: helm/oauth2-proxy/ci/servicemonitor-values.yaml [hunk 15]
```diff
@@ -1,4 +1,18 @@
 metrics:
   enabled: true
-  serviceMonitor:
+  servicemonitor:
```
Comment (pierluigilenoci) [inline helm/oauth2-proxy/ci/servicemonitor-values.yaml:3]: Please use the camel case serviceMonitor.

**Review by pierluigilenoci — COMMENTED** [review]
> I hope I hit all your points

Unfortunately not yet.

File: helm/oauth2-proxy/values.yaml (hunk 5 above)
Comment (jkroepke) [reply helm/oauth2-proxy/values.yaml]: Removed

**Comment by jkroepke** [comment]
After #171, `serviceMonitor` is valid now.

please review.

**Review by pierluigilenoci — APPROVED** [review]

**Pull request merged**